Promote "Convert to interpolated string" to an analyzer/fixer (#68469)

Bartleby2718 opened the pull request:
This is my first time looking into a `dotnet` repository or working with VisualBasic & XUnit, so I figured I should do the skeleton first and then move on to the next step. (I expect to go through a LOT of iterations.)

In my first commit, I created all the files I think I'd need. I think I understand the overall structure of the `src/Analyzers` directory.

For the first iteration, can I get some feedback on the following?
1. whether I created all files (e.g. do I need a test for VisualBasic?)
2. whether the code matches the convention of this repository (e.g. when should I use named parameters?)

Also, could you assign the GitHub issue and the pull request to me, so that it's easier for me to find them?

Fixes #68469

CyrusNajmabadi commented:
> (e.g. do I need a test for VisualBasic?)

I haven't looked closely, but my recollection is that the refactoring works for VB and C#.  So i'd expect the analyzer to do the same.  In which case we would want tests.

> (e.g. when should I use named parameters?)

Generally when passing literals that are non-bvious in context.  like `x.Remove(document, true); // what does true mean here?)`

CyrusNajmabadi commented:
> Also, could you assign the GitHub issue

Done.

CyrusNajmabadi commented:
> and the pull request to me, 

Not needed.  you can just click on here in github: 

![image](https://github.com/dotnet/roslyn/assets/4564579/7516721f-ade0-4e3a-b896-877ca2c2c955)

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/ConvertToInterpolatedString/ConvertToInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,31 @@
+namespace Microsoft.CodeAnalysis.CSharp.ConvertToInterpolatedString;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal abstract class ConvertToInterpolatedStringDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
```
this should likely derive from a common analyzer for both VB and C#.  It can derive from AbstractBuiltInCodeStyleDiagnosticAnalyzer

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/ConvertToInterpolatedString/ConvertToInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,31 @@
+        : base(
+            IDEDiagnosticIds.ConvertToInterpolatedStringDiagnosticId,
+            EnforceOnBuildValues.ConvertToInterpolatedString,
+            option: null,
```
will likely need an option controllin this.

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/CSharpCodeFixesResources.resx:
```diff
@@ -227,6 +227,9 @@
     <value>Remove 'new' modifier</value>
   </data>
+  <data name="Convert_to_interpolated_string" xml:space="preserve">
+    <value>Convert to interpolated string</value>
```
should be at the common layer snice VB would use the same string.

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/ConvertToInterpolatedString/ConvertToInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,33 @@
+namespace Microsoft.CodeAnalysis.CSharp.ConvertToInterpolatedString;
+
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.ConvertToInterpolatedString), Shared]
+internal partial class ConvertToInterpolatedStringCodeFixProvider : AbstractConvertToInterpolatedStringCodeFixProvider
```
likely could be in common layer.  will dpeend if the fix provider itself needs to do something specific for c#.

Bartleby2718 replied:
@CyrusNajmabadi I don't know anything about VB, but this doesn't seem too hard to refactor later, so I suppose I can just keep this conversation open and resolve later?

CyrusNajmabadi commented on src/Analyzers/CSharp/Tests/ConvertToInterpolatedString/ConvertToInterpolatedStringTests.cs:
```diff
@@ -0,0 +1,45 @@
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.ConvertToInterpolatedString
```
nice!

please use file-scoped namespaces thought.

CyrusNajmabadi commented on src/Analyzers/Core/Analyzers/ConvertToInterpolatedString/AbstractConvertToInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace Microsoft.CodeAnalysis.ConvertToInterpolatedString;
+
+internal abstract class AbstractConvertToInterpolatedStringDiagnosticAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
```
C# analyzer should derive from this.

CyrusNajmabadi commented on src/Analyzers/Core/Analyzers/EnforceOnBuildValues.cs:
```diff
@@ -96,6 +96,7 @@ internal static class EnforceOnBuildValues
     public const EnforceOnBuild UseCollectionExpressionForCreate = /*IDE0303*/ EnforceOnBuild.Recommended;
     public const EnforceOnBuild UseCollectionExpressionForBuilder = /*IDE0304*/ EnforceOnBuild.Recommended;
     public const EnforceOnBuild UseCollectionExpressionForFluent = /*IDE0305*/ EnforceOnBuild.Recommended;
+    public const EnforceOnBuild ConvertToInterpolatedString = /*IDE0310*/ EnforceOnBuild.Recommended;
```
Can you make this 0320.  i'd like teh whole 0300-19 rance for collection expressions.  thanks!

CyrusNajmabadi commented on src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -198,6 +198,7 @@ internal static class IDEDiagnosticIds
     public const string UseCollectionExpressionForCreateDiagnosticId = "IDE0303";
     public const string UseCollectionExpressionForBuilderDiagnosticId = "IDE0304";
     public const string UseCollectionExpressionForFluentDiagnosticId = "IDE0305";
+    public const string ConvertToInterpolatedStringDiagnosticId = "IDE0310";
```
These should be called `UseInterpolatedString`.  `Convert` is hte action to get yo uto the desired state.  But the diagnostic id states the desired style preference.  In this case, it's to use interpolated strings.

CyrusNajmabadi replied:
Note; we will likely end up wanting separate diagnostic ids for each pattern we detect.  Stephen listed several.  this will make it possible for users to enable/disable the diagnostic for some cases htey don't wnat .

Bartleby2718 replied:
Will keep this conversation open to remind myself about multiple patterns.

CyrusNajmabadi commented:
Converted this to draft.  This looks *great*.  I think it was a very good idea to start with the scaffolding first.  Note: if you run into issues https://discord.com/channels/143867839282020352/598678594750775301 is good for help (often 'live').

Bartleby2718 commented:
@CyrusNajmabadi Thanks for the quick review! A few more questions:
1. For some reason, I can't join the Discord server just by clicking it (either on desktop or on mobile). Is there an invite link?
2. What should I do when I want another round of review--just click `Re-request review`? 
![image](https://github.com/dotnet/roslyn/assets/7676354/05f2d12e-20c9-4327-a96c-c8b970aae71c)
3. Should I write tests for both analyzer and codefixprovider?

CyrusNajmabadi commented:
> What should I do when I want another round of review--just click Re-request review?

yes.  

Note: i try to get to them quickly, but it may take a couple of days.  Feel free to ping for a review (though try not to ping more htan once a day) :)

> Discord issues

Checking on that.

> Should I write tests for both analyzer and codefixprovider?

yes... but since we're moving the refactoring over to be an analyzer/fixprovider, you should be able to move all the tests over.

note: you don't need to test the analyzer in isolation.  you can just write tests that validate both at the same time for a given input.

CyrusNajmabadi commented:
For discord, use https://discord.com/invite/tGJvv88.  Sorry about that!

Bartleby2718 commented:
No big changes so far, but I've added the option for `IDE0320` and renamed a bunch of things. Since I'm in no rush, asking for another round of review @CyrusNajmabadi. If this looks good (e.g. class names, option name, resource name/description), I'll start moving things from `ConvertRegularStringToInterpolatedStringRefactoringProvider.cs` to the `Analyzer`/`CodeFixProvider`.

CyrusNajmabadi commented on src/Analyzers/CSharp/Analyzers/UseInterpolatedString/CSharpUseInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,27 @@
+namespace Microsoft.CodeAnalysis.CSharp.UseInterpolatedString;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal abstract class CSharpUseInterpolatedStringDiagnosticAnalyzer : AbstractUseInterpolatedStringDiagnosticAnalyzer
```
why abstract?

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,33 @@
+{
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpUseInterpolatedStringCodeFixProvider()
```
nit: can use primary constructors.

Note: `nits` are minor suggestions.  You are free to ignore them if you don't want to do them :)

CyrusNajmabadi replied:
(just say you're ignoring it :))

Bartleby2718 replied:
VS didn't seem to have an analyzer for this, so I had to fix this manually. Not sure if this is worth reporting, but I'll keep this thread open for now.

CyrusNajmabadi commented on src/Analyzers/Core/CodeFixes/UseInterpolatedString/AbstractUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Microsoft.CodeAnalysis.UseInterpolatedString;
+
+internal abstract partial class AbstractUseInterpolatedStringCodeFixProvider : CodeFixProvider
```
shoudl likely derive from SyntaxEditorBasedCodeFixProvider

CyrusNajmabadi commented on src/Analyzers/Core/CodeFixes/UseInterpolatedString/AbstractUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,19 @@
+
+internal abstract partial class AbstractUseInterpolatedStringCodeFixProvider : CodeFixProvider
+{
+    public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
```
should likely not use the batch fixer.

CyrusNajmabadi commented:
Looks fantastic!

DoctorKrolic commented on src/Analyzers/Core/Analyzers/IDEDiagnosticIds.cs:
```diff
@@ -198,6 +198,7 @@ internal static class IDEDiagnosticIds
     public const string UseCollectionExpressionForCreateDiagnosticId = "IDE0303";
     public const string UseCollectionExpressionForBuilderDiagnosticId = "IDE0304";
     public const string UseCollectionExpressionForFluentDiagnosticId = "IDE0305";
+    public const string UseInterpolatedStringDiagnosticId = "IDE0330";
 
     public const string MakeAnonymousFunctionStaticDiagnosticId = "IDE0320";
```
```suggestion

    public const string MakeAnonymousFunctionStaticDiagnosticId = "IDE0320";
    
    public const string UseInterpolatedStringDiagnosticId = "IDE0330";
```

Bartleby2718 replied:
Fixed in https://github.com/dotnet/roslyn/pull/73008/commits/b586235e2b496d486d5888c46b60b7fe245a1275

Bartleby2718 commented on src/Analyzers/CSharp/Tests/UseInterpolatedString/CSharpUseInterpolatedStringTests.cs:
```diff
@@ -0,0 +1,245 @@
+    }.RunAsync();
+
+    [Fact]
+    public Task FixSingleLineRawStrings() => new VerifyCS.Test
```
I got some tests to pass, but this test and the next one are failing for a reason I cannot understand.
![image](https://github.com/dotnet/roslyn/assets/7676354/5d756d2b-4b45-4943-a05d-0285781beb9e)
Even though I'm using the right start/end tokens in https://github.com/dotnet/roslyn/pull/73008/files#diff-db1e05fa98ce7807a912e3f653c71fb03ce8fb5eb7e96d0e39b585f0858cceeaR66-R87
I'm not seeing that in the `FixedCode`. Any idea why this would be?

CyrusNajmabadi replied:
i see the issue.  i'll comemnt accordingly.

Bartleby2718 commented:
I still have a lot more work to do, but requesting a review as I have no idea why these tests are failing.

CC: @CyrusNajmabadi

CyrusNajmabadi commented:
I can take a look tomorrow.  Thanks :)

CyrusNajmabadi commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+    private static SyntaxKind GetEndToken(SyntaxKind kind) => kind switch
+    {
+        SyntaxKind.SingleLineRawStringLiteralToken => SyntaxKind.InterpolatedRawStringEndToken,
+        SyntaxKind.MultiLineRawStringLiteralToken => SyntaxKind.InterpolatedRawStringEndToken,
```
so the problem here is that's not sufficient to just return a kind here.  Basically, many tokens in teh language have a 1:1 mapping between their kind and the exact text that must appear in the token contents.  For example the ClassKeyword kind exactly maps to `class`.  This doesn't apply to `InterpolatedRawStringEndToken` because the end token contents are *variable*.  For example, a raw string can end with `"""` or `""""` or `"""""` or any number of quotes.  Similarly, an interplated raw start token could be `$"""` `$""""` `$"""""` and so on.  It can also be `$$"""` `$$$"""` and so on.  Basically it can start with some number of dollars, at least 3 quotes, and then must end with the same number of quotes that it started with.

Bartleby2718 replied:
@CyrusNajmabadi
1. In that case, should I manually find out the start token (whether it be three double quotes or four double quotes or more) and have the same number of double quotes in the end token?
2. Once I determine that, would the next step be `SyntaxFactory.ParseToken("\"\"\"\"")`?

CyrusNajmabadi replied:
1. Yes. You'll need to figure it out. 
2.  When creating a token, you can specify the text content. You don't need to parse it. 

Thanks!

Bartleby2718 replied:
@CyrusNajmabadi  Taking a step back, I see in https://github.com/dotnet/roslyn/blob/main/src%2FFeatures%2FCore%2FPortable%2FConvertToInterpolatedString%2FConvertRegularStringToInterpolatedStringRefactoringProvider.cs#L44-L45 that raw string literals are not touched by the existing refactoring. Should I maintain this behavior?

Context: I was debugging raw string literal tests 
![image](https://github.com/dotnet/roslyn/assets/7676354/83eb4110-996b-44c0-86aa-891884586609)
and realized these tests didn't exist in https://github.com/dotnet/roslyn/blob/0d1f9fea04892d9e65e9a780bb773a4d5ed18f8e/src/Features/CSharpTest/ConvertToInterpolatedString/ConvertRegularStringToInterpolatedStringTests.cs.

CyrusNajmabadi replied:
> that raw string literals are not touched by the existing refactoring. Should I maintain this behavior?

Up to you.  I didn't do it because there wasn't really the need for such an upgrade path.

Put another way, when the language started, we didn't have interpolated strings.  So the idea of these features was to take the older patterns people would use to make strings, and move them to use the nicer new stuff.

But raw-strings *are* hte nice new stuff.  And when we did raw strings, we made sure we had normal raw strings and interpolated raw strings at the same time.  So there's a lot less need to go "oh... someone built a complex expr with raw strings, that they now want to convert to an interpolated form".  Instead, generally, people just start with the interpolated form in the first place.

Bartleby2718 replied:
Okay, I guess I misunderstood the intent of the refactoring then. I thought this feature would be useful when you wanted to convert a regular string to an interpolated string because you now want to use variables in it, as in [here](https://github.com/nunit/nunit.analyzers/pull/721/files#diff-4eb6ac27ac8fd046156ed20ecb78685c3f0f8a330f3a88e6084462494026d098R140-R145). Since that can be true for raw string literals, I was going to do the same for them.

If it's up to me, I think I'll do a little more of poking around to see if I can do this. It didn't seem straightforward on a cursory look.

DoctorKrolic commented on src/Analyzers/CSharp/Analyzers/UseInterpolatedString/CSharpUseInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,58 @@
+            return false;
+
+        // If there is a const keyword, do not offer the refactoring
+        // An interpolated string is not const
```
"An interpolated string is not const" - that is not true. C# 10 added [constant interpolated strings](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/constant_interpolated_strings). In short, if all interpolation variables are constants, the interpolation is considered a constant as well, e.g.:
```cs
const string a = "a";
const string b = $"Value of 'a' is: {a}";
```
Whether to support this case is up to you. If this requires just several touches, then go ahead. If not, feel free to leave it as is. But in either way the comment should be either removed or changed to say that supporting constant interpolations would require too much effort or something similar

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseInterpolatedString), Shared]
+[method: ImportingConstructor]
+[method: SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
```
Given that you've already ported tests to `VerifyCS` you should now be able to remove suppression:
```suggestion
[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
```
P.S. Don't just blindly commit this suggestion, it might require an appropriate `using` to be added. Copy the new attribute and paste it to the code via VS, it should figure out and add that `using` automatically

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+[ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseInterpolatedString), Shared]
+[method: ImportingConstructor]
+[method: SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+internal partial class CSharpUseInterpolatedStringCodeFixProvider() : AbstractUseInterpolatedStringCodeFixProvider
```
```suggestion
internal sealed partial class CSharpUseInterpolatedStringCodeFixProvider() : AbstractUseInterpolatedStringCodeFixProvider
```

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+    {
+        var span = diagnostic.Location.SourceSpan;
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var node = root.FindNode(span, getInnermostNodeForTie: true) as LiteralExpressionSyntax;
+        if (node?.Span != span || !node.IsKind(SyntaxKind.StringLiteralExpression))
```
Better to have a strongly typed node:
```suggestion
        var node = root.FindNode(span, getInnermostNodeForTie: true) as StringLiteralExpressionSyntax;
        if (node?.Span != span)
```

Bartleby2718 replied:
@DoctorKrolic Is there such a thing? I couldn't find it.
- https://github.com/search?q=repo%3Adotnet%2Froslyn%20StringLiteralExpressionSyntax&type=code
- https://www.google.com/search?q=%22StringLiteralExpressionSyntax%22

Should I create one?

DoctorKrolic replied:
Yep, my bad. I thought there is a corresponding syntax node to every kind of literal, which is wrong, there is only `LiteralExpressionSyntax`. So just resolve ang ignore this feedback)

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+        editor.ReplaceNode(node, newNode);
+    }
+
+    private static InterpolatedStringExpressionSyntax CreateInterpolatedString(SyntaxNode stringLiteralExpressionNode)
```
Given my suggestion above you should be able to strongly type parameter node here as well:
```suggestion
    private static InterpolatedStringExpressionSyntax CreateInterpolatedString(StringLiteralExpressionSyntax stringLiteralExpressionNode)
```

DoctorKrolic replied:
I see, you've already changed this parameter type to `LiteralExpressionSyntax`, good. Feedback resolved

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+                text: text,
+                valueText: text,
+                trailing: SyntaxTriviaList.Empty))
+            .WithTriviaFrom(stringLiteralExpressionNode);
```
Isn't there an overload of `InterpolatedStringText`, that accepts leading and trailing trivia, so we can avoid unnecessary intermidiate node allocation?

Bartleby2718 replied:
Nope!
![image](https://github.com/dotnet/roslyn/assets/7676354/f1ba77a1-3dcd-4c78-8847-6fa3f810ff19)

DoctorKrolic commented on src/Analyzers/CSharp/CodeFixes/UseInterpolatedString/CSharpUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,103 @@
+    };
+
+    private static int GetStartIndex(SyntaxKind kind, bool isVerbatimString) => (isVerbatimString ? 1 : 0)
+        + kind switch
```
I find these 2 lines hard to read. Given that this is not a trivial switch I would recommend using block body here and _not_ put operands of an addition on different lines. Maybe for clarify even use separate local for storing a result of the switch and the return `(isVerbatimString ? 1 : 0) + newLocal`

DoctorKrolic commented on src/Analyzers/CSharp/Tests/UseInterpolatedString/CSharpUseInterpolatedStringTests.cs:
```diff
@@ -0,0 +1,245 @@
+public sealed class CSharpUseInterpolatedStringTests
+{
+    [Fact]
+    public Task KeepInterpolatedStrings() => new VerifyCS.Test
```
Suggestion №1: You can use `VerifyCS.VerifyCodeFixAsync(codem fixedCode)` when you just need to specify input and output without additional configurations.
Suggestion №2: In roslyn codebase tests like this usually have block bodies and actually `await` the result of the test rather than return the task from it. Please visit other tests, see how things are done there and be consistent with the rest of the codebase, thanks!

DoctorKrolic commented on src/Analyzers/CSharp/Tests/UseInterpolatedString/CSharpUseInterpolatedStringTests.cs:
```diff
@@ -0,0 +1,245 @@
+    public Task KeepInterpolatedStrings() => new VerifyCS.Test
+    {
+        TestCode = """
+        class Program
```
Raw string should be idented 1 more level from property name in this and all remaining test cases. I guess you've copied it from somewhere and formatter didn't force-ident the raw string, it isn't that aggressive with their indentation

DoctorKrolic commented on src/Analyzers/CSharp/Tests/UseInterpolatedString/CSharpUseInterpolatedStringTests.cs:
```diff
@@ -0,0 +1,245 @@
+        ExpectedDiagnostics =
+        {
+            // /0/Test0.cs(3,33): hidden IDE0330: String can be converted to interpolated string
+            VerifyCS.Diagnostic().WithSpan(3, 33, 3, 43),
```
Please use test markup for diagnostic locations. Just surround the location with `[|` and `|]` and remove these manually specified `ExpectedDiagnostics`, this should do the trick

DoctorKrolic commented on src/Analyzers/Core/Analyzers/UseInterpolatedString/AbstractUseInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,66 @@
+
+internal abstract class AbstractUseInterpolatedStringDiagnosticAnalyzer<TSyntaxKind, TExpressionSyntax, TStringLiteralExpressionSyntax>
+    : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    where TSyntaxKind : struct
```
Can this be constrained more specificly?
```suggestion
    where TSyntaxKind : struct, Enum
```

DoctorKrolic commented on src/Analyzers/Core/Analyzers/UseInterpolatedString/AbstractUseInterpolatedStringDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+    }
+
+    protected abstract bool CanConvertToInterpolatedString(TStringLiteralExpressionSyntax stringLiteralExpression, ParseOptions parseOptions);
```
Please move this abstract declaration somewhere to the top of the declaration. From my experience, it should be placed even before constructor, so it is immidiatelly obvious, what abstract things does this class have when reading from top to bottom. Maybe visit other similar places to verify, that mys suggested location is consistent with what is already there in the codebase

DoctorKrolic commented on src/Analyzers/Core/CodeFixes/UseInterpolatedString/AbstractUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Microsoft.CodeAnalysis.UseInterpolatedString;
+
+internal abstract partial class AbstractUseInterpolatedStringCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```
This doesn't need to be `partial`, right?
```suggestion
internal abstract class AbstractUseInterpolatedStringCodeFixProvider : SyntaxEditorBasedCodeFixProvider
```

DoctorKrolic commented on src/Analyzers/Core/CodeFixes/UseInterpolatedString/AbstractUseInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+    }
+
+    protected abstract Task FixOneAsync(Document document, Diagnostic diagnostic, SyntaxEditor editor, CancellationToken cancellationToken);
```
Same comment with the location of an abstract member declaration as in the `AbstractUseInterpolatedStringDiagnosticAnalyzer` class

DoctorKrolic commented on src/Analyzers/VisualBasic/CodeFixes/UseInterpolatedString/VisualBasicUseInterpolatedStringCodeFixProvider.vb:
```diff
@@ -0,0 +1,37 @@
+        Inherits AbstractUseInterpolatedStringCodeFixProvider
+
+        <ImportingConstructor>
+        <SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification:="Used in test code: https://github.com/dotnet/roslyn/issues/42814")>
```
A friendly reminder that this should not stay suppressed when you are done with VB part

Bartleby2718 replied:
Thanks! Left a TODO comment for now. Should be easy to find before merge.

DoctorKrolic commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/CodeStyle/IdeCodeStyleOptions.cs:
```diff
@@ -46,6 +46,7 @@ internal record class IdeCodeStyleOptions
     [DataMember] public CodeStyleOption2<bool> PreferConditionalExpressionOverAssignment { get; init; } = CodeStyleOption2.TrueWithSilentEnforcement;
     [DataMember] public CodeStyleOption2<bool> PreferConditionalExpressionOverReturn { get; init; } = CodeStyleOption2.TrueWithSilentEnforcement;
     [DataMember] public CodeStyleOption2<bool> PreferCompoundAssignment { get; init; } = CodeStyleOption2.TrueWithSuggestionEnforcement;
+    [DataMember] public CodeStyleOption2<bool> PreferInterpolatedString { get; init; } = CodeStyleOption2.TrueWithSuggestionEnforcement;
```
This should be initialized in the constructor below. Please keep it after the `PreferCompoundAssignment` there as well
